core/src/main/java/com/exadel/etoolbox/rolloutmanager/core/services/impl/PageReplicationServiceImpl.java

```diff
@@ -0,0 +1,118 @@
+
+@Component(service = PageReplicationService.class)
+@Designate(ocd = PageReplicationServiceImpl.Configuration.class)
+public class PageReplicationServiceImpl implements PageReplicationService {
+    private static final Logger LOG = LoggerFactory.getLogger(PageReplicationServiceImpl.class);
+
+    @ObjectClassDefinition(name = "EToolbox Page Replication Service Configuration")
+    @interface Configuration {
+
+        @AttributeDefinition(
+                name = "Pool size",
+                description = "The number of Threads in the pool")
+        int poolSize() default 5;
+    }
+
+    @Activate
+    private PageReplicationServiceImpl.Configuration config;
+
+    @Reference
+    private LiveRelationshipManager liveRelationshipManager;
+
+    @Reference
+    private JobManager jobManager;
+
+    @Reference
+    private Replicator replicator;
+
+    public List<RolloutStatus> replicateItems(ResourceResolver resourceResolver, RolloutItem[] items, PageManager pageManager) {
+        return Arrays.stream(items)
+                .collect(Collectors.groupingBy(RolloutItem::getDepth))
+                .entrySet()
+                .stream()
+                .sorted(Map.Entry.comparingByKey())
+                .map(Map.Entry::getValue)
+                .flatMap(sortedByDepthItems -> replicateSortedByDepthItems(resourceResolver, sortedByDepthItems, pageManager))
+                .collect(Collectors.toList());
+    }
+
+    private Stream<RolloutStatus> replicateSortedByDepthItems(ResourceResolver resourceResolver, List<RolloutItem> items, PageManager pageManager) {
+        ExecutorService executorService = Executors.newFixedThreadPool(config.poolSize());
+        return items.stream()
+                .filter(item -> StringUtils.isNotBlank(item.getTarget()))
+                .map(item -> CompletableFuture.supplyAsync(() -> replicate(resourceResolver, item, pageManager), executorService))
+                .collect(Collectors.toList())
+                .stream()
+                .map(CompletableFuture::join);
+    }
+
+    private RolloutStatus replicate(ResourceResolver resourceResolver, RolloutItem targetItem, PageManager pageManager) {
+
+        String targetPath = targetItem.getTarget();
+        RolloutStatus status = new RolloutStatus(targetPath);
+
+        Optional<Page> targetPage = Optional.ofNullable(pageManager.getPage(targetPath));
+        Session session = resourceResolver.adaptTo(Session.class);
+        if (!targetPage.isPresent() || ObjectUtils.isEmpty(session)) {
+            status.setSuccess(false);
+            LOG.warn("Replication failed - target page is null, page path: {}", targetPath);
+            return status;
+        }
+        try {
+            replicator.replicate(session, ReplicationActionType.ACTIVATE, targetPath);
+        } catch (ReplicationException ex) {
+            status.setSuccess(false);
+            LOG.error("Exception during page replication", ex);
+        }
+        status.setSuccess(true);
```

Comment:
❗ Need to move this inside `try`, or else the status success will be reset to "true" after set to "false"

Reply:
done